src/libraries/System.Private.CoreLib/src/System/DateTime.cs

```diff
@@ -249,79 +315,467 @@ public DateTime(int year, int month, int day, int hour, int minute, int second,
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DateTime"/> structure to the specified year, month, day, hour, minute, second,
+        /// millisecond, and Coordinated Universal Time (UTC) or local time for the specified calendar.
+        /// </summary>
+        /// <param name="year">The year (1 through 9999).</param>
+        /// <param name="month">The month (1 through 12).</param>
+        /// <param name="day">The day (1 through the number of days in <paramref name="month"/>).</param>
+        /// <param name="hour">The hours (0 through 23).</param>
+        /// <param name="minute">The minutes (0 through 59).</param>
+        /// <param name="second">The seconds (0 through 59).</param>
+        /// <param name="millisecond">The milliseconds (0 through 999).</param>
+        /// <param name="microsecond">The milliseconds (0 through 900).</param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="year"/> is less than 1 or greater than 9999.
+        ///
+        /// -or-
+        ///
+        /// <paramref name="month"/> is less than 1 or greater than 12.
+        ///
+        /// -or-
+        ///
+        /// <paramref name="day"/> is less than 1 or greater than the number of days in <paramref name="month"/>.
+        ///
+        /// -or-
+        ///
+        /// <paramref name="hour"/> is less than 0 or greater than 23.
+        ///
+        /// -or-
+        ///
+        /// <paramref name="minute"/> is less than 0 or greater than 59.
+        ///
+        /// -or-
+        ///
+        /// <paramref name="second"/> is less than 0 or greater than 59.
+        ///
+        /// -or-
+        ///
+        /// <paramref name="millisecond"/> is less than 0 or greater than 999.
+        ///
+        /// -or-
+        ///
+        /// <paramref name="microsecond"/> is less than 0 or greater than 900.
+        /// </exception>
+        /// <remarks>
+        /// This constructor interprets <paramref name="year"/>, <paramref name="month"/> and <paramref name="day"/> as a year, month and day
+        /// in the Gregorian calendar. To instantiate a <see cref="DateTime"/> value by using the year, month and day in another calendar, call
+        /// the <see cref="DateTime(int, int, int, int, int, int, int, int, Calendar)"/> constructor.
+        ///
+        /// The <see cref="Kind"/> property is initialized to <see cref="DateTimeKind.Unspecified"/>.
+        ///
+        /// For applications in which portability of date and time data or a limited degree of time zone awareness is important,
+        /// you can use the corresponding <see cref="DateTimeOffset"/> constructor.
+        /// </remarks>
+        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microsecond)
         {
             if ((uint)millisecond >= MillisPerSecond) ThrowMillisecondOutOfRange();
```

Comment:
per my other comment, I expect the code here can be factored out without any change and be used in this old constructor and the new one.

Reply:
We cannot factor it out into a new method, because this requires updating `_dateData` which is `readonly` and only possible ctor.

Also we cannot reassign `this` because `DateTime` is a `readonly` struct.

Reply:
Here are some rough implementation demonstrates how you can do it. The constructor taking a calendar may need some more tweaks.

```C#
        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, DateTimeKind kind) => _dateData = Init(year, month, day, hour, minute, second, millisecond, kind);

        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microssecond, DateTimeKind kind)
        {
            ulong ticks = Init(year, month, day, hour, minute, second, millisecond, kind);
            if ((uint)microsecond >= MicrosecondsPerMillisecond) ThrowMicrosecondOutOfRange();

            ulong newTicks = (ticks & TicksMask) + (microsecond * TicksPerMicrosecond);
            
            // Maybe the following condition is not really needed? adding microseconds shouldn't exceed MaxTicks anyway here.
            if (newTicks > MaxTicks) ThrowDateArithmetic(0);
            _dateData = newTicks | (ticks & FlagsMask);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal ulong Init(int year, int month, int day, int hour, int minute, int second, int millisecond, DateTimeKind kind)
        {
            if ((uint)millisecond >= MillisPerSecond) ThrowMillisecondOutOfRange();
            if ((uint)kind > (uint)DateTimeKind.Local) ThrowInvalidKind();

            if (second != 60 || !s_systemSupportsLeapSeconds)
            {
                ulong ticks = DateToTicks(year, month, day) + TimeToTicks(hour, minute, second);
                ticks += (uint)millisecond * (uint)TicksPerMillisecond;
                Debug.Assert(ticks <= MaxTicks, "Input parameters validated already");
                return ticks | ((ulong)kind << KindShift);
            }
            else
            {
                // if we have a leap second, then we adjust it to 59 so that DateTime will consider it the last in the specified minute.
                DateTime dt = new DateTime(year, month, day, hour, minute, 59, millisecond, kind);

                if (!IsValidTimeWithLeapSeconds(year, month, day, hour, 59, kind))
                {
                    ThrowHelper.ThrowArgumentOutOfRange_BadHourMinuteSecond();
                }

                return dt._dateData;
            }
        }

```